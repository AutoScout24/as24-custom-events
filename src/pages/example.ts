// Events used for template integration / testing. Do not remove

/**
 * Fictitious fragment initialization event
 */
export const SomeFragmentInitializationExample = 'SOME_FRAGMENT:INITIALIZED'
export type SomeFragmentInitializationExample = {
  /**
   * Example payload
   */
  foo: boolean
}

/**
 * Tatsu service template integration example event
 */
export const TatsuServiceTemplateSelectImage = 'TATSU_SERVICE_TEMPLATE:SELECT_IMAGE'
export type TatsuServiceTemplateSelectImage = {
  /**
   * Example payload
   */
  selectedImage: number
}


// Add events to Document
declare global {
  interface DocumentEventMap {
    [SomeFragmentInitializationExample]: CustomEvent<SomeFragmentInitializationExample>
    [TatsuServiceTemplateSelectImage]: CustomEvent<TatsuServiceTemplateSelectImage>
  }
}
