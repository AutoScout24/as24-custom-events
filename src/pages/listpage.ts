export const ClassifiedListFilterUpdate = 'CL_FILTER_UPDATE'
export type ClassifiedListFilterUpdate = {
  searchUrls: {
    /**
     * Search url with legacy parameters
     */
    legacy: string
    /**
     * Search url with newer `mmm` parameters
     */
    standard: string
    /**
     * Standard search query without page parameter
     */
    unpaginatedStandard: string
    /**
     * Full search url including path
     */
    fullUrl: string
  }
}

export const WebPushOptIn = 'WEB_PUSH:OPT_IN'
export type WebPushOptIn = {
  /**
   * id of the customer who opt-in
   */
  as24Visitor: string
  /**
   * FCM push token
   */
  pushToken: string
}

export const ClassifiedListTotalCountUpdate = 'TOTAL_COUNT_UPDATE'
export type ClassifiedListTotalCountUpdate = {
  totalCount: number | null
}

/**
 * Used by the list-item summary fragment to bind internal events
 */
export const ListItemsChanged = 'list-items:changed'
export type ListItemsChanged = undefined

/**
 * Indicates the list items summary fragment has been wired internally
 */
export const ListItemsInitialized = 'LIST_ITEMS_INITIALIZED'
export type ListItemsInitialized = undefined

/**
 * Indicates the list items summary fragment has been wired internally
 */
export const SearchIdUpdated = 'SEARCH_ID_UPDATED'
export type SearchIdUpdated = undefined

export const SavedSearchesFragmentInitialization = 'SAVED_SEARCHES_FRAGMENT_INIT'
export type SavedSearchesFragmentInitialization = undefined

/**
 * Retrieve the current filters. The list page will fire a CL_FILTER_UPDATE as a response
 */
export const ClassifiedListRequestFiltersValue = 'CLASSIFIED_LIST:REQUEST_FILTER_VALUES'
export type ClassifiedListRequestFiltersValue = undefined

/**
 * Indicate when filter edit fragment has been initialized
 */
export const ClassifiedListFilterEditInitialized = 'CLASSIFIED_LIST:FILTER_EDIT_INITIALIZED'
export type ClassifiedListFilterEditInitialized = undefined

// inform compiler of the as24 custom events
declare global {
  interface DocumentEventMap {
    [ClassifiedListFilterUpdate]: CustomEvent<ClassifiedListFilterUpdate>
    [ClassifiedListRequestFiltersValue]: CustomEvent<ClassifiedListRequestFiltersValue>
    [ClassifiedListTotalCountUpdate]: CustomEvent<ClassifiedListTotalCountUpdate>
    [SavedSearchesFragmentInitialization]: CustomEvent<SavedSearchesFragmentInitialization>
    [ListItemsChanged]: CustomEvent<ListItemsChanged>
    [ListItemsInitialized]: CustomEvent<ListItemsInitialized>
    [SearchIdUpdated]: CustomEvent<SearchIdUpdated>
    [WebPushOptIn]: CustomEvent<WebPushOptIn>
    [ClassifiedListFilterEditInitialized]: CustomEvent<ClassifiedListFilterEditInitialized>
  }
}
