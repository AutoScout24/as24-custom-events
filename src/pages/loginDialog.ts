export const LoginDialogEvent = 'AS24_LOGIN_DIALOG_EVENT'
export const LoginDialogTypeChange = 'AS24_LOGIN_DIALOG_CHANGE'

export enum LoginDialogType {
    NONE = 'NONE',
    SIGN_IN_AND_UP = 'SIGN_IN_AND_UP',
    SIGN_IN = 'SIGN_IN',
    SIGN_UP = 'SIGN_UP',
}

export type LoginDialogTypeChange = {
    type: typeof LoginDialogTypeChange
    dialog: LoginDialogType
}

export type LoginDialogEvent = LoginDialogTypeChange

declare global {
    interface DocumentEventMap {
        [LoginDialogEvent]: CustomEvent<LoginDialogEvent>
        [LoginDialogTypeChange]: CustomEvent<LoginDialogTypeChange>
    }
}