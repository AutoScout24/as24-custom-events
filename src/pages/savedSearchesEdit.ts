export { ClassifiedListFilterEditInitialized, ClassifiedListFilterUpdate } from './listpage'

/**
 * Update search mask query
 */
export const SearchMaskFilterUpdate = 'SAVED_SEARCHES_EDIT:SEARCH_MASK_QUERY_UPDATE'
export type SearchMaskFilterUpdate = {
  query: string
}

declare global {
  interface DocumentEventMap {
    [SearchMaskFilterUpdate]: CustomEvent<SearchMaskFilterUpdate>
  }
}
