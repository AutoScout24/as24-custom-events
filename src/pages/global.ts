import * as ListPage from './listpage'

// Events that are present across all pages

/**
 * Watchlist event
 */
export const WatchlistEvent = 'AS24_WATCHLIST_EVENT'
export const WatchlistEventCountUpdate = 'WATCHLIST_COUNT_UPDATE'
export type WatchlistEventCountUpdate = {
  type: typeof WatchlistEventCountUpdate
  value: number
}
export type WatchlistEvent = WatchlistEventCountUpdate

// inform compiler of the as24 custom events
declare global {
  interface DocumentEventMap {
    [WatchlistEvent]: CustomEvent<WatchlistEvent>
  }
}
