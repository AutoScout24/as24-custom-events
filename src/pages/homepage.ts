import * as ListPage from './listpage'

export const TestEvent = 'HOMEPAGE_TEST_EVENT'
export type TestEvent = {
  foo: boolean
}

// Add events to Document
declare global {
  interface DocumentEventMap {
    [TestEvent]: CustomEvent<TestEvent>
  }
}
