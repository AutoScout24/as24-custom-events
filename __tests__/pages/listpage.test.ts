import * as p from '../../src/pages/listpage'

describe('listpage events', () => {
  const eventNames = [
    p.ClassifiedListFilterUpdate,
    p.ClassifiedListTotalCountUpdate,
    p.ListItemsChanged,
    p.SavedSearchesFragmentInitialization,
    p.ListItemsInitialized,
    p.WebPushOptIn,
    p.ClassifiedListRequestFiltersValue
  ]

  test.each(eventNames)('event names must never change', eventName => {
    expect(eventName).toMatchSnapshot()
  })

  test('type definitions are inferred correctly by compiler', () => {
    document.addEventListener(p.ClassifiedListFilterUpdate, e => e.detail.searchUrls.legacy)
    document.addEventListener(p.ClassifiedListTotalCountUpdate, e => e.detail.totalCount)
    document.addEventListener(p.ListItemsChanged, e => e.detail as undefined)
    document.addEventListener(p.ListItemsInitialized, e => e.detail as undefined)
    document.addEventListener(p.SavedSearchesFragmentInitialization, e => e.detail as undefined)
    document.addEventListener(p.WebPushOptIn, e => e.detail.as24Visitor as string)
    document.addEventListener(p.ClassifiedListRequestFiltersValue, e => e.detail as undefined)
    expect(true).toBeTruthy() // dummy assertion since compiler check is the real test
  })
})
