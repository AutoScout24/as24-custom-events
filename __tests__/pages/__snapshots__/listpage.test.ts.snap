// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`listpage events event names must never change 1`] = `"CL_FILTER_UPDATE"`;

exports[`listpage events event names must never change 2`] = `"TOTAL_COUNT_UPDATE"`;

exports[`listpage events event names must never change 3`] = `"list-items:changed"`;

exports[`listpage events event names must never change 4`] = `"SAVED_SEARCHES_FRAGMENT_INIT"`;

exports[`listpage events event names must never change 5`] = `"LIST_ITEMS_INITIALIZED"`;

exports[`listpage events event names must never change 6`] = `"WEB_PUSH:OPT_IN"`;

exports[`listpage events event names must never change 7`] = `"CLASSIFIED_LIST:REQUEST_FILTER_VALUES"`;
